File: app/src/main/java/com/babestudios/companyinfouk/ui/insolvency/InsolvencyActivity.java
package com.babestudios.companyinfouk.ui.insolvency;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.babestudios.companyinfouk.CompaniesHouseApplication;
import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.insolvency.Insolvency;
import com.babestudios.companyinfouk.data.model.insolvency.InsolvencyCase;
import com.babestudios.companyinfouk.ui.insolvencydetails.InsolvencyDetailsActivity;
import com.babestudios.companyinfouk.utils.DividerItemDecoration;
import com.google.gson.Gson;

import net.grandcentrix.thirtyinch.TiActivity;

import javax.inject.Inject;
import javax.inject.Singleton;

import butterknife.Bind;
import butterknife.ButterKnife;

public class InsolvencyActivity extends TiActivity<InsolvencyPresenter, InsolvencyActivityView> implements InsolvencyActivityView, InsolvencyAdapter.InsolvencyRecyclerViewClickListener {

	@Bind(R.id.toolbar)
	Toolbar toolbar;

	@Bind(R.id.insolvency_recycler_view)
	RecyclerView insolvencyRecyclerView;

	@Bind(R.id.lblNoInsolvency)
	TextView lblNoInsolvency;

	private InsolvencyAdapter insolvencyAdapter;

	@Bind(R.id.progressbar)
	ProgressBar progressbar;

	@Singleton
	@Inject
	DataManager dataManager;

	String companyNumber;

	@SuppressWarnings("ConstantConditions")
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_insolvency);

		ButterKnife.bind(this);
		if (toolbar != null) {
			setSupportActionBar(toolbar);
			getSupportActionBar().setDisplayHomeAsUpEnabled(true);
			getSupportActionBar().setTitle(R.string.insolvency);
			toolbar.setNavigationOnClickListener(v -> onBackPressed());
		}
		companyNumber = getIntent().getStringExtra("companyNumber");
		createInsolvencyRecyclerView();


	}
	private void createInsolvencyRecyclerView() {
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
		insolvencyRecyclerView.setLayoutManager(linearLayoutManager);
		insolvencyRecyclerView.addItemDecoration(
				new DividerItemDecoration(this));

		/*insolvencyRecyclerView.addOnScrollListener(new EndlessRecyclerViewScrollListener(linearLayoutManager) {
			@Override
			public void onLoadMore(int page, int totalItemsCount) {
				getPresenter().loadMoreCharges(page);
			}
		});*/
	}

	@Override
	public void showProgress() {
		progressbar.setVisibility(View.VISIBLE);
	}

	@Override
	public void hideProgress() {
		progressbar.setVisibility(View.GONE);
	}


	@Override
	public void showInsolvency(Insolvency insolvency) {
		if (insolvencyRecyclerView.getAdapter() == null) {
			insolvencyAdapter = new InsolvencyAdapter(InsolvencyActivity.this, insolvency, dataManager);
			insolvencyRecyclerView.setAdapter(insolvencyAdapter);
		} else {
			insolvencyAdapter.updateItems(insolvency);
		}
	}

	@NonNull
	@Override
	public InsolvencyPresenter providePresenter() {
		CompaniesHouseApplication.getInstance().getApplicationComponent().inject(this);
		return new InsolvencyPresenter(dataManager);
	}

	@Override
	public void insolvencyItemClicked(View v, int position, InsolvencyCase insolvencyCase) {
		Gson gson = new Gson();
		String jsonItem = gson.toJson(insolvencyCase, InsolvencyCase.class);
		Intent intent = new Intent(this, InsolvencyDetailsActivity.class);
		intent.putExtra("insolvencyCase", jsonItem);
		startActivity(intent);
	}

	@Override
	public String getCompanyNumber() {
		return companyNumber;
	}

	@Override
	public void showNoInsolvency() {
		lblNoInsolvency.setVisibility(View.VISIBLE);
		insolvencyRecyclerView.setVisibility(View.GONE);
	}

	@Override
	public void showError() {
		Toast.makeText(this, "Error", Toast.LENGTH_LONG).show();
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/insolvency/InsolvencyActivity.kt
package com.babestudios.companyinfouk.ui.insolvency

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.babestudios.base.mvp.ErrorType
import com.babestudios.base.mvp.list.BaseViewHolder
import com.babestudios.base.view.DividerItemDecoration
import com.babestudios.base.view.MultiStateView.*
import com.babestudios.companyinfo.core.injection.CoreInjectHelper
import com.babestudios.companyinfouk.R
import com.babestudios.companyinfouk.ext.logScreenView
import com.babestudios.companyinfouk.ext.startActivityWithRightSlide
import com.babestudios.companyinfouk.ui.insolvency.list.AbstractInsolvencyVisitable
import com.babestudios.companyinfouk.ui.insolvency.list.InsolvencyAdapter
import com.babestudios.companyinfouk.ui.insolvency.list.InsolvencyTypeFactory
import com.babestudios.companyinfouk.ui.insolvency.list.InsolvencyViewHolder
import com.babestudios.companyinfouk.ui.insolvencydetails.createInsolvencyDetailsIntent
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.ScopeProvider
import com.ubercab.autodispose.rxlifecycle.RxLifecycleInterop
import io.reactivex.CompletableSource
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_insolvency.*
import kotlinx.android.synthetic.main.multi_state_view_error.view.*

private const val COMPANY = "com.babestudios.companyinfouk.ui.company"

class InsolvencyActivity : RxAppCompatActivity(), ScopeProvider {


	private var insolvencyAdapter: InsolvencyAdapter? = null

	override fun requestScope(): CompletableSource = RxLifecycleInterop.from(this).requestScope()

	private val viewModel by lazy { ViewModelProviders.of(this).get(InsolvencyViewModel::class.java) }

	private lateinit var insolvencyPresenter: InsolvencyPresenterContract

	private val eventDisposables: CompositeDisposable = CompositeDisposable()

	private lateinit var comp: InsolvencyComponent

	//region life cycle

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_insolvency)
		logScreenView(this.localClassName)
		setSupportActionBar(pabInsolvency.getToolbar())
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		pabInsolvency.setNavigationOnClickListener { onBackPressed() }
		supportActionBar?.setTitle(R.string.insolvency)
		when {
			viewModel.state.value.insolvencyItems != null -> {
				initPresenter(viewModel)
			}
			savedInstanceState != null -> {
				savedInstanceState.getParcelable<InsolvencyState>("STATE")?.let {
					with(viewModel.state.value) {
						insolvencyItems = it.insolvencyItems
						companyNumber = it.companyNumber
					}
				}
				initPresenter(viewModel)
			}
			else -> {
				viewModel.state.value.companyNumber = intent.getStringExtra(COMPANY)
				initPresenter(viewModel)
			}
		}

		createRecyclerView()
		observeState()
	}

	override fun onResume() {
		super.onResume()
		observeActions()
	}

	override fun onSaveInstanceState(outState: Bundle) {
		outState.putParcelable("STATE", viewModel.state.value)
		super.onSaveInstanceState(outState)
	}

	private fun initPresenter(viewModel: InsolvencyViewModel) {
		if (!::insolvencyPresenter.isInitialized) {
			comp = DaggerInsolvencyComponent
					.builder()
					.coreComponent(CoreInjectHelper.provideCoreComponent(applicationContext))
					.build()
			insolvencyPresenter = comp.insolvencyPresenter()
			insolvencyPresenter.setViewModel(viewModel, requestScope())
		}
	}

	private fun createRecyclerView() {
		val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
		rvInsolvency?.layoutManager = linearLayoutManager
		rvInsolvency.addItemDecoration(DividerItemDecoration(this))
	}

	override fun onBackPressed() {
		super.finish()
		overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out)
	}

	//endregion

	//region render

	private fun observeState() {
		viewModel.state
				.`as`(AutoDispose.autoDisposable(this))
				.subscribe { render(it) }
	}

	private fun render(state: InsolvencyState) {
		when {
			state.isLoading -> msvInsolvency.viewState = VIEW_STATE_LOADING
			state.errorType != ErrorType.NONE -> {
				msvInsolvency.viewState = VIEW_STATE_ERROR
				state.errorType = ErrorType.NONE
				msvInsolvency.tvMsvError.text = state.errorMessage
			}
			state.insolvencyItems == null -> msvInsolvency.viewState = VIEW_STATE_EMPTY
			else -> {
				state.insolvencyItems?.let {
					msvInsolvency.viewState = VIEW_STATE_CONTENT
					if (rvInsolvency?.adapter == null) {
						insolvencyAdapter = InsolvencyAdapter(it, InsolvencyTypeFactory())
						rvInsolvency?.adapter = insolvencyAdapter
					} else {
						insolvencyAdapter?.updateItems(it)
					}
					observeActions()
				}
			}
		}
	}

	//endregion

	//region events

	private fun observeActions() {
		eventDisposables.clear()
		insolvencyAdapter?.getViewClickedObservable()
				?.take(1)
				?.`as`(AutoDispose.autoDisposable(this))
				?.subscribe { view: BaseViewHolder<AbstractInsolvencyVisitable> ->
					viewModel.state.value.insolvencyItems?.let { insolvencyItems ->
						startActivityWithRightSlide(
								this.createInsolvencyDetailsIntent(insolvencyItems[(view as InsolvencyViewHolder).adapterPosition].insolvencyCase))
					}
				}
				?.let { eventDisposables.add(it) }
	}

	//endregion
}

fun Context.createInsolvencyIntent(company: String): Intent {
	return Intent(this, InsolvencyActivity::class.java)
			.putExtra(COMPANY, company)
}
